designs/vpce_api_flexibility.md

```diff
@@ -0,0 +1,337 @@
+      record:
+        hostname: "splunk"
+        externalNameService:
+          name: "examplesvcname"
+```
+
+```mermaid
+graph LR
+  B[VPCE] --> C
+  subgraph Splunk Hosting Account
+  C[Splunk VPCE Service]
+  end
+  subgraph Customer Account
+  subgraph ROSA Cluster
+  A[ExternalName Service]
+  end
+  A --Cluster's Private Hosted Zone--> B
+  end
+```
+
+### Planned FedRAMP OCM Use-Case
+
+Creating a VPC Endpoint in the same account and VPC as the cluster. VPC Endpoint resolution via private DNS configured
+by the VPC Endpoint Service and no custom DNS configuration.
+
+```yaml
+apiVersion: avo.openshift.io/v1alpha2
+kind: VpcEndpoint
+metadata:
+  name: demo
+  namespace: demo
+spec:
+  enablePrivateDns: true
+  serviceName: "com.amazonaws.vpce.us-east-1.vpce-svc-00000000000000000"
+  securityGroup:
+    ingressRules:
+      - fromPort: 443
+        toPort: 443
+        protocol: "tcp"
+  vpc:
+    useRosaPrivateSubnets: true
+```
+
+```mermaid
+graph LR
+  B[VPCE w/ Private DNS] --> C
+  subgraph OCM Hosting Account
+  C[VPCE Service w/ Private DNS]
+  end
+  subgraph Customer Account
+  D[ROSA Cluster] --> B
+  end
+```
+
+Creating a VPC Endpoint in the same account and VPC as the cluster. Custom DNS resolution via a separate Route 53
+Private Hosted Zone created by AVO.
+
+```yaml
+apiVersion: avo.openshift.io/v1alpha2
+kind: VpcEndpoint
+metadata:
+  name: demo
+  namespace: demo
+spec:
+  serviceName: "com.amazonaws.vpce.us-east-1.vpce-svc-00000000000000000"
+  securityGroup:
+    ingressRules:
+      - fromPort: 443
+        toPort: 443
+        protocol: "tcp"
+  vpc:
+    useRosaPrivateSubnets: true
+  customDns:
+    route53PrivateHostedZone:
+      useRosaPrivateHostedZone: false
+      domainName: "redhat.com"
+      record:
+        hostname: "ocm"
+        externalNameService:
+          name: "ocm"
+```
+
+```mermaid
+graph LR
+  B[VPCE] --> C
+  subgraph OCM Hosting Account
+  C[VPCE Service]
+  end
+  subgraph Customer Account
+  subgraph ROSA Cluster
+  A[ExternalName Service]
+  end
+  A --OCM Private Hosted Zone--> B
+  end
+```
+
+### Planned HyperShift Use-Case
+
+Creating a Private Hosted Zone and VPC Endpoint in a different account, and potentially a different region with a
+Hosted Zone record in a preexisting Route 53 Private Hosted Zone.
+
+```yaml
+apiVersion: avo.openshift.io/v1alpha2
+kind: VpcEndpoint
+metadata:
+  name: demo
+  namespace: demo
+spec:
+  assumeRoleArn: "arn:aws:iam::12345678910:role/OtherArn"
+  region: "us-west-1"
+  serviceName: "com.amazonaws.vpce.us-west-1.vpce-svc-00000000000000000"
+  securityGroup:
+    ingressRules:
+      - fromPort: 443
+        toPort: 443
+        protocol: "tcp"
+  vpc:
+    useRosaPrivateSubnets: false
+    subnetIds:
+      - "subnet-1"
+      - "subnet-2"
+      - "subnet-3"
+  customDns:
+    route53PrivateHostedZone:
+      useRosaPrivateHostedZone: false
+      id: "HOSTEDZONEID"
+      record:
+        hostname: "some.hostname"
```

Comment:
The hostname would not be an FQDN, right? Do we want to change this to FQDN?

Reply:
Yeah, this would not be a FQDN. I'll fix this example